### ISSUE TITLE
`Dns.GetHostAddressesAsync` roots diagnostic types even if they are disabled (trimming)

### ISSUE THREAD

**Issue by neon-sunset** [dotnet/runtime#101355]
### Description
It appears that calling `Dns.GetHostAddressesAsync` roots diagnostic types that are unused by accessing `IsEnabled` flag within `NameResolutionMetrics` which is implemented in the way that is not analyzable by the trimmer.

![image](https://github.com/dotnet/runtime/assets/20912188/1f6daf25-f82d-4d07-a647-32e28ecf3f40)

### Environment
```
.NET SDK:
 Version:           8.0.300-preview.24203.14 (no idea why this is preview)
 Commit:            b7e38f457d
 Workload version:  8.0.300-manifests.c4df6daf
 MSBuild version:   17.10.0+4f6b1bb28

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.22631
 OS Platform: Windows
 RID:         win-x64
 Base Path:   C:\Program Files\dotnet\sdk\8.0.300-preview.24203.14\
 ```
 
### Data

Please find the mstat file from a [repro](https://github.com/U8String/U8String/tree/main/Examples/TwitchChatReader) attached: [TwitchChatReader.mstat.zip](https://github.com/dotnet/runtime/files/15053322/TwitchChatReader.mstat.zip)

### Analysis

It was not immediately apparent if there exists a way to let the linker/trimmer know that no diagnostics are expected. Perhaps I'm missing something?

Thanks.

**Comment by MihaZupan** [comment]
Are you using the `System.Diagnostics.Metrics.Meter.IsSupported` feature switch to disable the functionality?

**Comment by neon-sunset** [comment]
(if you are asking me, then I can't - it is `internal`, there is no code that references `System.Diagnostics.Metrics` namespace directly)

**Comment by MihaZupan** [comment]
I'm talking about https://github.com/dotnet/runtime/blob/main/docs/workflow/trimming/feature-switches.md#available-feature-switches

**Comment by neon-sunset** [comment]
I was unaware of it, thanks! (missed the link at the end of [this](https://learn.microsoft.com/en-us/dotnet/core/deploying/trimming/trimming-options?pivots=dotnet-8-0#trimming-framework-library-features) section) Setting `<MetricsSupport>false</MetricsSupport>` does indeed reduce the binary size but the types still end up being rooted:
![image](https://github.com/dotnet/runtime/assets/20912188/f6423103-b0b1-4ea5-85fa-c6c18ec8d726)

Diff of removed:
![image](https://github.com/dotnet/runtime/assets/20912188/ce3ef89b-03da-495e-9508-356e2f97d27b)

As a side note, is there a reason this is enabled by default for default SDK template publishes as AOT that does not reference `EventListener`? (or does NativeAOT support event pipe now?)

**Comment by MichalStrehovsky** [comment]
@MihaZupan is `MetricsSupport` a publicly supported switch? If so, could you or someone please add it to the list at https://learn.microsoft.com/en-us/dotnet/core/deploying/trimming/trimming-options?pivots=dotnet-6-0#trimming-framework-library-features (you can click the "Edit this document" at the top of the page to edit it on Github).

Are metrics usable with `EventSourceSupport=false`? If not, we might want to consider `MetricsSupport` as false if `EventSourceSupport` is false.

> or does NativeAOT support event pipe now?

It supports it, but by default we set EventSourceSupport to false when PublishAot is true. Some SDKs (like the ASP.NET SDK) will change the default to true.

Cc @eerhardt

**Comment by eerhardt** [comment]
> is MetricsSupport a publicly supported switch?

Yes, it was introduced for Xamarin Android. See https://github.com/dotnet/runtime/pull/91767.

> Are metrics usable with EventSourceSupport=false?

Yes, I believe so. One example is using OpenTelemetry to export the metrics using say OTLP, or prometheus. EventSource shouldn't be needed in those scenarios.

**Comment by tarekgh** [comment]
I changed the area label as the metrics work correctly and `Dns.GetHostAddressesAsync` is what root the types. No action needed from metrics side. I understand AOT need to do some work here and they can triage this issue accordingly.
